include/RAJA/exec-agency/forall_agency.hxx

```diff
@@ -0,0 +1,145 @@
+//   this list of conditions and the disclaimer (as noted below) in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of the LLNS/LLNL nor the names of its contributors may
+//   be used to endorse or promote products derived from this software without
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/int_datatypes.hxx"
+#include "RAJA/fault_tolerance.hxx"
+#include "RAJA/segment_exec.hxx"
+#include "RAJA/ThreadUtils_CPU.hxx"
+
+#include "agency/agency.hpp"
+#include "agency/experimental.hpp"
+
+#include <thread>
+#include <iterator>
+#include <algorithm>
+
+namespace RAJA
+{
+
+/// ASSUMPTIONS:
+///   1- That we can get the number of threads this way
+///   2- That the number of workers can be sensibly computed this way
+///   3- That the number of workers computed this way works for 
+///      sequenctuial execution.
+template <typename Func, typename Agent, typename Worker>
+RAJA_INLINE void forall(const agency_base<Agent, Worker>&, 
+                        const RangeSegment& iter, 
+                        Func&& loop_body)
+{
+  auto numThreads = getMaxReduceThreadsCPU();
+  auto tiles = agency::experimental::tile_evenly(
+      agency::experimental::interval(iter.getBegin(), iter.getEnd()), numThreads);
+
+  agency::bulk_invoke(Worker{}(tiles.size()),
+                      [=](Agent& self) {
+                          for (Index_type i : tiles[self.index()]) {
+                              loop_body(i);
+                          }
+                      });
+}
+
+template <typename Iterable, typename Func, typename Agent, typename Worker>
+RAJA_INLINE void forall(const agency_base<Agent, Worker>&,
+                        Iterable&& iter,
+                        Func&& loop_body)
+{
+  auto numThreads = getMaxReduceThreadsCPU();
+  auto distance = std::end(iter) - std::begin(iter);
+
+  auto tiles = agency::experimental::tile_evenly(
+      agency::experimental::interval(0, distance), numThreads);
```

Comment:
Same deal, except we tile up the size of our range, as it doesn't natively support iterators.